### ISSUE TITLE
Sentry Virtual Filesystem v2 (VFS2)

### ISSUE THREAD

**Issue by fvoznika** [google/gvisor#1035]
Current VFS implementation in the Sentry has performance shortcomings with regarding to the number of operations that need to be performed for path walks. This is especially noticeable when using gofer mounted filesystems, where the round trip cost for each operation is aggravated by the RPC and scheduling costs.

VFS2 addresses this by delegating path resolution to the filesystem, making it possible to send a single RPC for each syscall operation, instead of one RPC per path component in the operation. For example, `stat(/foo/bar/goo)` generates at least 3 RPC round trips to the gofer (`foo`, `bar`, `goo`), while VFS2 makes only 1.

VFS2 also allows for the implementation of recursive bind mounts, which would require a major refactoring with the current implementation.

**Comment by qobilidop** [comment]
Hi @fvoznika, we are a team of 3 graduate students (@Reapor-Yurnero @JFHwang @qobilidop) [looking for a gVisor performance-related project to satisfy a course requirement](https://groups.google.com/forum/#!topic/gvisor-users/g68bG0x0AO8). It seems there is a lot of work to be done for this filesystem improvement project. We wonder if there is a chance for us to participate in a small part of this. If it's okay, where should we get started? Thanks a lot in advance!

**Comment by fvoznika** [comment]
VFS2 is feature complete. Vast majority were converted to run with VFS2 enabled and they are all passing. We're starting to enable VFS2 internally at Google and will flip the default to VFS2 once we have soaked it in production for some time.

We recommend everyone to start running with the `--vfs2` flag to catch any possible issue early on, before we switch over to VFS2 by default.

**Comment by iangudger** [comment]
@fvoznika Is VFS2 ready to be the default? The last update was almost a year ago.

**Comment by fvoznika** [comment]
Yes, it's ready to be the default. We're just waiting for a few internal users to enable it before we flip the flag. But everyone is encouraged to run with `--vfs2` at this point.
